designs/0005-static-matrices.md

```diff
@@ -0,0 +1,343 @@
+
+2. A design for exposing these autodiff variables at the Stan language level
+
+An autodiff type that allows customization of the underlying storage object makes it possible
+for matrices, instead of being data structures of autodiff types, for the autodiff type
+itself to represent a matrix. Operations on these new types can be more efficient than the old
+ones, and this design remains fully backwards compatible with the curren Stan autodiff types.
+
+For the rest of this proposal, "autodiff" means "reverse mode autodiff," and all C++ functions
+are assumed to reside in the `stan::math` namespace. The necessary namespace will be
+included if this is not true.
+
+# New Reverse Mode Autodiff Type
+[newtype]: #newtype
+
+The new autodiff type is `stan::math::var_value<T>`, where `T` is the underlying
+storage type of the autodiff variable. For `T == double`, this type is
+equivalent to the current `stan::math::var` type.
+
+A `stan::math::var` in Stan is a PIMPL (pointer to implementation) object. It is
+a pointer to a `stan::math::vari`, which is an object that holds the actual
+value and adjoint. In terms of data, the basic var and vari implementation look
+like:
+
+```cpp
+class vari {
+  double value_;
+  double adjoint_;
+};
+
+class var {
+  vari* vi_; // pointer to implementation
+}
+```
+
+The new, templated `var_value<T>` and `vari_value<T>` are very similar except
+the `double` type is replaced with a template type `T`:
+
+```cpp
+template <typename T>
+class vari {
+  T value_;
+  T adjoint_;
+};
+
+template <typename T>
+class var_value<T> {
+  vari_value<T>* vi_; // pointer to implementation
+}
+```
+
+Currently, Stan implements matrices using
+`Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>` variables. Internally,
+an `N x M` `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>` is a
+pointer to `N x M` vars on the heap. Because vars are themselves just pointers
+to varis, the current Stan matrix class is represented internally as a pointer
+to pointers.
+
+When Stan performs an operation on a matrix it will typically need
+to read all of the values of the matrix at least once on the forward autodiff
+pass and then update all the adjoints at least once on the reverse autodiff
+pass. This can be slow because the pointer chasing required to read a pointer to
+pointers data structure is expensive. The can be wasteful because fetching just
+the values or adjoints ends up loading both the values and adjoints into cache
+(since they are stored beside each other). Stan vectors and row vectors are
+implemented similarly and present the same problems.
+
+This proposal will enable an `N x M` Stan matrix to be defined in terms of one
+autodiff object with two separate `N x M` containers, one for its values and
+one for its adjoints. Filling in the templates, the storage of a
+`var_value<Eigen::MatrixXd>` is effectively:
+
+```cpp
+template <>
+class vari_value<Eigen::MatrixXd> {
+  Eigen::MatrixXd value_;
+  Eigen::MatrixXd adjoint_;
+}
```

Comment:
I'd put this back to the way I wrote it because on a first glance this code is going to leave dangling memory in `Eigen::MatrixXd`. I would go the opposite and just explain that we can use allocate the memory in our stack allocator but use `Eigen::Map<>` to act as if these are normal Eigen matrices

Reply:
I'm gonna add a note in t he comments. It won't actually leak memory cause it's not being allocated in the `arena` according to this code either :D.

The difficulty of C++ psuedo-code is it's basically guaranteed to be wrong (unless it's a copy of the implementation). Vague pseudo-code is nicer cause you have wiggle room.

But ofc. this discussion came up the first round too: https://github.com/stan-dev/design-docs/pull/21#discussion_r421806784

And I see Bob made comments that it was sort of inappropriate to just link C++ implementations like I've done here.